File: python/lesson 16/find_median_from_data_stream.py
# https://leetcode.com/problems/find-median-from-data-stream/

class MedianFinder:

    def __init__(self):
        self.maxHeap = []
        self.minHeap = []

    def addNum(self, num):
        if len(self.maxHeap) == len(self.minHeap):
            heappush(self.minHeap, -heappushpop(self.maxHeap, -num))
        else:
            heappush(self.maxHeap, -heappushpop(self.minHeap, num))

    def findMedian(self):
        if len(self.maxHeap) == len(self.minHeap):
            return float(self.minHeap[0] - self.maxHeap[0]) / 2.0
        else:
            return float(self.minHeap[0])

File: csharp/lesson 16/FindMedianFromDataStream.cs
// https://leetcode.com/problems/find-median-from-data-stream/

public class MedianFinder {
    private PriorityQueue<int, int> maxHeap = new PriorityQueue<int, int>(Comparer<int>.Create((a, b) => b.CompareTo(a)));
    private PriorityQueue<int, int> minHeap = new PriorityQueue<int, int>();
    
    
    public void AddNum(int num) {
        if (maxHeap.Count == minHeap.Count) {
            minHeap.Enqueue(num, num);
            maxHeap.Enqueue(minHeap.Peek(), minHeap.Dequeue());
        } else {
            maxHeap.Enqueue(num, num);
            minHeap.Enqueue(maxHeap.Peek(), maxHeap.Dequeue());            
        }
    }
    
    public double FindMedian() {
         return maxHeap.Count == minHeap.Count
            ? (minHeap.Peek() + maxHeap.Peek()) / 2.0
            : maxHeap.Peek();
    }
}
